pytorch_lightning/utilities/fetching.py

```diff
@@ -236,19 +234,27 @@ def prefetching(self) -> None:
                 break
 
     def fetching_function(self) -> Tuple[Any, bool]:
+        assert self.dataloader_iter is not None
         if self.batches:
+            # we pre-fetched, consume one
             batch = self.batches.pop(0)
-        else:
-            # empty iterator, no prefetching done
-            raise StopIteration
-        if not self.done:
-            assert self.dataloader_iter is not None
             try:
+                # and replace it
                 self._fetch_next_batch(self.dataloader_iter)
             except StopIteration:
+                self.done = not self.batches
+        elif not self.done:
+            # we did not prefetch at all
+            try:
+                self._fetch_next_batch(self.dataloader_iter)
```

Comment:
This comment is confusing as it is not fully accurate. You might have actually pre-fetched.

Reply:
If prefetching was done (in `prefetching`) then the `if self.batches` block above will be the one to always run as it will also replace the batch consumed.

This block is an elif so it will not execute when the above block does.